src/audio/host-zephyr.c

```diff
@@ -378,9 +377,7 @@ static uint32_t host_get_copy_bytes_normal(struct host_data *hd, struct comp_dev
 		return 0;
 	}
 
-	dma_buf_c = buffer_acquire(hd->dma_buffer);
-	dma_sample_bytes = get_sample_bytes(audio_stream_get_frm_fmt(&dma_buf_c->stream));
-	buffer_release(dma_buf_c);
+	dma_sample_bytes = hd->config.src_width;
```

Comment:
ok, good - this is pre-calculated so can be reused on every iteration instead of looked up.

Reply:
in principle I think it's good to have this change, but in single core configurations impact should be minimal. In such configurations `buffer_acquire()` is a conversion from an uncached to a cached alias, `buffer_release()` is a NOP, and `get_sample_bytes()` is a primitive switch. So I really wouldn't expect this change to bring 5% of performance. However - if asserts are enabled - then yes, buffer acquisition will perform several more checks. Are you measuring performance with asserts enabled?

Reply:
it is debug build with -d option, so I would assume assert enabled.
BTW, I checked release build performance, it is much better than debug
                  host          dai
debug         396          455
release        340          405

we are seeing a nearly 15% difference between release and debug, it maybe caused by the assert.

Reply:
That is still surprisingly high savings. Looking at assembly, the buffer_acquire() variant does not do this much more work. There is more code that is jumped over in single core case, but this should affect runtime cost this much. Perhaps this is leading to additional instruction cache miss. It would be good to understand this better, so we could pick good/bad usages better in code review.

Reply:
for debug, I did not check the assembly code for buffer_acquire().
There is so much macro inside.
after patch: #8106 merge, I guess there is a performance improve for each component and whole pipeline.
hard to give a number, maybe 10%

Reply:
> release 340 405

this means, that with a release build (no debugging overlay, asserts disabled) before your patch(es) DSP load way 405 and after your patch(es) 340? Is it with all of them? I'm particularly wondering about commits 1 and 2, I really don't think we should go down to that level of optimisation. I think we can talk about 3 code "classes"
1. configuration. All the code outside the LL scheduler threads, i.e. for components everything outside `.copy()` / `.process()` calls. That isn't very performance critical, emphasis on correctness and clarity
2. inside .copy()` / `.process()` but outside "byte shuffling" loops - that should be reasonably optimised but I wouldn't go crazy there. Avoid wasteful code, but keep it clean and readable
3. the hottest spots - actual byte processing loops. These should be optimised to the maximum. Here we can sacrifice readability to squeeze out an additional instruction saving

So, I think in some of your commits you might be overdoing with "class 2" optimisations. Maybe we could put a threshold - if a specific "class 2" optimisation brings less than 5% "overall" (i.e. measured over the entire `.copy()` call) performance saving and reduces clarity, readability, introduces code reuse etc. - maybe we shouldn't take it.

Reply:
host      dai
debug build:  396      455
release build: 340      405

above table does not related this patch, it is just with same code base build with release and debug, different cycle cost for release and debug, it is obvious that release is faster 15-20% percent compared with debug.

I did not over-eng code, all the change are in critical timing path(1ms calling),  this function is mainly used perform host copy.